projects/flow.groups-2D-random-walk/src/project.py

```diff
@@ -0,0 +1,211 @@
+def compute_squared_distance(job):
+    """Compute the squared distance for a random walk."""
+    with job.data:
+        positions = job.data.positions[:]
+    job.data.squared_distance = np.sum(positions * positions, axis=1)
+
+
+# This operation happens after computing the msd so it isn't in base.
+# Also we use pre as a filter for jobs ensuring this job only ever runs on the zeroth
+# replica for a given standard deviation.
+@RandomWalkProject.pre.true("msd_analyzed")
+@RandomWalkProject.pre(lambda job: job.sp.replica == 0)
+@RandomWalkProject.post.isfile("msd.png")
+@RandomWalkProject.operation
+def plot_msd(job):
+    """Plot the MSD for all std."""
+    with job.data:
+        msd = job.data.msd[:]
+    fig, ax = plt.subplots()
+    ax.plot(msd)
+    ax.set_title(f"MSD for {job.sp.standard_deviation} std")
+    ax.set_xlabel("x")
+    ax.set_ylabel("MSD")
+    # Only save figure to the first replica
+    fig.savefig(job.fn("msd.png"))
+    fig.close()
+
+
+# Aggregate operations on individual standard deviations across replicas
+
+# Create aggregator that combines all replicas with a single standard deviation
+std_aggregator = flow.aggregator.groupby("standard_deviation")
+
+
+# Define all aggregate groups
+agg_plot = RandomWalkProject.make_group(
+    "aggregate-plot", group_aggregator=std_aggregator
+)
+agg_analyze_and_plot = RandomWalkProject.make_group(
+    "post-processing", group_aggregator=std_aggregator
+)
+
+
+@agg_analyze_and_plot
+@std_aggregator
+@RandomWalkProject.pre(all_simulated)
+@RandomWalkProject.post.true("msd_analyzed")
+@RandomWalkProject.operation
+def compute_mean_squared_distance(*jobs):
+    """Compute and store the mean squared displacement for all std."""
+    squared_distance_iterator = generate_stores(jobs, "squared_distance")
+    squared_distance = next(squared_distance_iterator)
+    msd = np.copy(squared_distance)
+    for squared_distance in squared_distance_iterator:
+        msd += squared_distance
+    msd /= len(jobs)
+    # Store msd in only first replica (job.sp.replica == 1)
+    job_replica_zero = get_replica(jobs, replica=0)
+    job_replica_zero.data["msd"] = msd
+    for job in jobs:
+        job.doc.msd_analyzed = True
+
+
+# Since this uses a separate aggragator to restrict aggregates to the first 5 replicas,
+# we cannot assign this operation to either agg_plot or agg_analyze_and_plot
+@flow.aggregator.groupby("standard_deviation", select=lambda job: job.sp.replica <= 4)
+@RandomWalkProject.pre(all_simulated)
+@RandomWalkProject.post.true("plotted_walks")
+@RandomWalkProject.operation
+def plot_walk(*jobs):
+    """Plot the first 5 replicas random walks for each standard_deviation."""
+    fig, ax = plt.subplots()
+    for position, job in zip(generate_stores(jobs, "positions"), jobs):
+        ax.plot(position[:, 0], position[:, 1], label=f"Replica {job.sp.replica}")
+    ax.legend()
+    ax.set_title(
+        f"Random Walks with {jobs[0].sp.standard_deviation} standard_deviation"
```

Comment:
```suggestion
        f"Random Walks with Standard Deviation {jobs[0].sp.standard_deviation}"
```

Reply:
Part of me wishes that the aggregators could provide information like the grouping parameter (equal to `jobs[0].sp.standard_deviation`). Could it make sense to have functions with keyword-only arguments like this? Maybe there's a fancy way to make this keyword-only argument optional with function inspection in the aggregator's decorator...

```python
@flow.aggregator.groupby("standard_deviation", select=lambda job: job.sp.replica <= 4)
@RandomWalkProject.operation
def plot_walk(*jobs, groupby_value):
    assert jobs[0].sp.standard_deviation == groupby_value
```

Reply:
The problem is the shear flexibility of aggregating jobs. 

For `groupby` specifically:
 We could use inspection, but what happens with multiple `groupby` keys, do we support one keyword argument for each key or just return a `dict`? If the user specifies the keys in a different order than the keyword arguments do we pass the values by keyword name (if supporting multiple keyword only arguments)? If only one key is specified and we pass a `dict` for multiple keys specified, do we still pass a `dict` or just the value?

These are all addressable though, my preference would be to return a `dict` always with one keyword only argument.

However, in general though for aggregation using `aggregator_function` this is impossible or very difficult to do. My concern is providing this feature for one type of aggregation, but not any others.